app/pages/_mobx.tsx

```diff
@@ -0,0 +1,281 @@
+import { ExpandMore, ChevronRight } from "@mui/icons-material";
+import { TreeItem, TreeView } from "@mui/lab";
+import {
+  CircularProgress,
+  Button,
+  Typography,
+  Box,
+  TextField,
+} from "@mui/material";
+import keyBy from "lodash/keyBy";
+import {
+  observable,
+  makeObservable,
+  computed,
+  flow,
+  configure,
+  ObservableMap,
+} from "mobx";
+import { observer } from "mobx-react-lite";
+import { useMemo, useState } from "react";
+import { Client, OperationResult, useClient } from "urql";
+
+import { DataSource } from "@/configurator";
+import {
+  DataCubeMetadataWithComponentValuesDocument,
+  DataCubeMetadataWithComponentValuesQuery,
+  DataCubeMetadataWithComponentValuesQueryVariables,
+  Dimension,
+  DimensionHierarchyDocument,
+  DimensionHierarchyQuery,
+  DimensionHierarchyQueryVariables,
+  HierarchyValue,
+  Measure,
+} from "@/graphql/query-hooks";
+import { visitHierarchy } from "@/rdf/tree-utils";
+
+configure({ enforceActions: "always" });
+
+type QueryResult<T> = {
+  fetchStatus: "loaded" | "loading" | "idle";
+  error: Error | undefined;
+  data: T | undefined;
+};
+
+class Cube {
+  client: Client;
+  iri: string;
+  source: DataSource;
+  dimensions: undefined | Dimension[];
+  measures: undefined | Measure[];
+  loading: boolean;
+  hierarchies: ObservableMap<string, QueryResult<HierarchyValue[]>>;
+
+  constructor(client: Client, iri: string, source: DataSource) {
+    this.client = client;
+    this.iri = iri;
+    this.source = source;
+    this.loading = false;
+    this.dimensions = undefined;
+    this.measures = undefined;
+    this.hierarchies = observable.map<string, QueryResult<HierarchyValue[]>>();
+    makeObservable(this, {
+      loading: observable,
+      dimensions: observable.struct,
+      measures: observable.struct,
+      components: computed,
+      componentsByIri: computed,
+      hierarchies: observable,
+      hierarchyParents: computed,
+      load: flow,
+      loadHierarchy: flow,
+    });
+  }
+
+  *load() {
+    const vars: DataCubeMetadataWithComponentValuesQueryVariables = {
+      iri: this.iri,
+      locale: "en",
+      sourceType: this.source.type,
+      sourceUrl: this.source.url,
+    };
+
+    try {
+      this.loading = true;
+      const res: OperationResult<DataCubeMetadataWithComponentValuesQuery> =
+        yield this.client
+          .query<DataCubeMetadataWithComponentValuesQuery>(
+            DataCubeMetadataWithComponentValuesDocument,
+            vars
+          )
+          .toPromise();
```

Comment:
`*load()` + yielding a promise to avoid doing `await Cube.load()`?

Reply:
This is unfortunately the way to do [async `actions`](https://mobx.js.org/actions.html#using-flow-instead-of-async--await-) in mobx